# Day 11 Completion Note

## Decision Tree Training Summary

- CIC-IDS2018 dataset was loaded successfully.
- 78 numeric features were used as model inputs.
- An 80:20 stratified train/test split was used.
- Invalid `NaN` and infinite values were identified in the model inputs.
- Infinite values were converted to `NaN`.
- Missing values were filled using medians calculated from the training data.
- The same training-derived medians were applied to the test data.
- Decision Tree model was trained successfully.
- Test predictions were generated for 209,715 samples.

## Decision Tree Baseline Results

- Accuracy: 0.9999523162
- Weighted Precision: 0.9999523286
- Weighted Recall: 0.9999523162
- Weighted F1 Score: 0.9999523161

The trained Decision Tree model was saved as:

`models/decision_tree_baseline.joblib`

The saved model was successfully loaded and verified.

**Day 11 status: Completed**

In [17]:
# Verify the saved Decision Tree model

import os
import joblib

print("Model file exists:", os.path.exists(dt_model_path))

# Load the saved model
dt_loaded = joblib.load(dt_model_path)

print("Saved model loaded successfully.")
print("Model type:", type(dt_loaded).__name__)

Model file exists: True
Saved model loaded successfully.
Model type: DecisionTreeClassifier


In [16]:
import joblib
import os

# Create models directory if it does not exist
os.makedirs("../models", exist_ok=True)

# Save trained Decision Tree model
dt_model_path = "../models/decision_tree_baseline.joblib"
joblib.dump(dt_model, dt_model_path)

print("Decision Tree model saved successfully.")
print("Saved to:", dt_model_path)

Decision Tree model saved successfully.
Saved to: ../models/decision_tree_baseline.joblib


In [15]:
# Store Decision Tree baseline results

dt_results = {
    "Model": "Decision Tree",
    "Accuracy": dt_accuracy,
    "Precision": dt_precision,
    "Recall": dt_recall,
    "F1 Score": dt_f1
}

print(dt_results)

{'Model': 'Decision Tree', 'Accuracy': 0.999952316238705, 'Precision': 0.9999523285658237, 'Recall': 0.999952316238705, 'F1 Score': 0.9999523161430673}


In [14]:
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    classification_report
)

# Calculate evaluation metrics
dt_accuracy = accuracy_score(y_test, y_pred_dt)
dt_precision = precision_score(
    y_test, y_pred_dt, average="weighted", zero_division=0
)
dt_recall = recall_score(
    y_test, y_pred_dt, average="weighted", zero_division=0
)
dt_f1 = f1_score(
    y_test, y_pred_dt, average="weighted", zero_division=0
)

print("Decision Tree Evaluation")
print("------------------------")
print("Accuracy :", round(dt_accuracy, 4))
print("Precision:", round(dt_precision, 4))
print("Recall   :", round(dt_recall, 4))
print("F1 Score :", round(dt_f1, 4))

print("\nClassification Report:")
print(classification_report(y_test, y_pred_dt, zero_division=0))

Decision Tree Evaluation
------------------------
Accuracy : 1.0
Precision: 1.0
Recall   : 1.0
F1 Score : 1.0

Classification Report:
                precision    recall  f1-score   support

        Benign       1.00      1.00      1.00    133525
FTP-BruteForce       1.00      1.00      1.00     38672
SSH-Bruteforce       1.00      1.00      1.00     37518

      accuracy                           1.00    209715
     macro avg       1.00      1.00      1.00    209715
  weighted avg       1.00      1.00      1.00    209715



In [13]:
# Generate predictions on the test set

print("Generating Decision Tree predictions...")

y_pred_dt = dt_model.predict(X_test_clean)

print("Prediction completed successfully.")
print("Number of predictions:", len(y_pred_dt))

Generating Decision Tree predictions...
Prediction completed successfully.
Number of predictions: 209715


In [12]:
# Train the Decision Tree model using cleaned inputs

print("Starting Decision Tree training...")

dt_model.fit(X_train_clean, y_train)

print("Decision Tree training completed successfully.")

Starting Decision Tree training...
Decision Tree training completed successfully.


In [11]:
# Final verification of cleaned model inputs

train_nan = X_train_clean.isna().sum().sum()
test_nan = X_test_clean.isna().sum().sum()

train_inf = np.isinf(X_train_clean).sum().sum()
test_inf = np.isinf(X_test_clean).sum().sum()

print("X_train_clean NaN:", train_nan)
print("X_train_clean Inf:", train_inf)

print("\nX_test_clean NaN:", test_nan)
print("X_test_clean Inf:", test_inf)

print(
    "\nTraining input is finite:",
    np.isfinite(X_train_clean.to_numpy()).all()
)

print(
    "Testing input is finite:",
    np.isfinite(X_test_clean.to_numpy()).all()
)

X_train_clean NaN: 0
X_train_clean Inf: 0

X_test_clean NaN: 0
X_test_clean Inf: 0

Training input is finite: True
Testing input is finite: True


In [10]:
# Fill missing values using medians calculated from training data

X_train_clean = X_train_clean.fillna(train_medians)
X_test_clean = X_test_clean.fillna(train_medians)

print("Missing values in X_train_clean:",
      X_train_clean.isna().sum().sum())

print("Missing values in X_test_clean:",
      X_test_clean.isna().sum().sum())

Missing values in X_train_clean: 0
Missing values in X_test_clean: 0


In [9]:
# Calculate median values from training data only

train_medians = X_train_clean.median()

print("Training medians for affected features:")
print(train_medians[["Flow Byts/s", "Flow Pkts/s"]])

Training medians for affected features:
Flow Byts/s      85.840257
Flow Pkts/s    2047.082907
dtype: float64


In [8]:
# Convert infinite values to NaN
# This prepares invalid values for median imputation.

X_train_clean = X_train.copy()
X_test_clean = X_test.copy()

X_train_clean = X_train_clean.replace([np.inf, -np.inf], np.nan)
X_test_clean = X_test_clean.replace([np.inf, -np.inf], np.nan)

print("Infinite values in X_train_clean:",
      np.isinf(X_train_clean).sum().sum())

print("Infinite values in X_test_clean:",
      np.isinf(X_test_clean).sum().sum())

Infinite values in X_train_clean: 0
Infinite values in X_test_clean: 0


In [7]:
# Check NaN and infinite values in the affected features

affected_features = ["Flow Byts/s", "Flow Pkts/s"]

for feature in affected_features:
    train_nan = X_train[feature].isna().sum()
    test_nan = X_test[feature].isna().sum()
    
    train_inf = np.isinf(X_train[feature]).sum()
    test_inf = np.isinf(X_test[feature]).sum()
    
    print(f"\n{feature}")
    print("  Training NaN :", train_nan)
    print("  Testing NaN  :", test_nan)
    print("  Training Inf :", train_inf)
    print("  Testing Inf  :", test_inf)


Flow Byts/s
  Training NaN : 1801
  Testing NaN  : 476
  Training Inf : 1240
  Testing Inf  : 307

Flow Pkts/s
  Training NaN : 0
  Testing NaN  : 0
  Training Inf : 3041
  Testing Inf  : 783


In [6]:
# Identify features containing infinite values

train_inf_by_column = np.isinf(X_train).sum()
test_inf_by_column = np.isinf(X_test).sum()

print("Columns with infinite values in training data:")
print(train_inf_by_column[train_inf_by_column > 0])

print("\nColumns with infinite values in testing data:")
print(test_inf_by_column[test_inf_by_column > 0])

Columns with infinite values in training data:
Flow Byts/s    1240
Flow Pkts/s    3041
dtype: int64

Columns with infinite values in testing data:
Flow Byts/s    307
Flow Pkts/s    783
dtype: int64


In [5]:
import numpy as np

# Check infinite values in training and testing inputs
train_inf = np.isinf(X_train).sum().sum()
test_inf = np.isinf(X_test).sum().sum()

print("Infinite values in X_train:", train_inf)
print("Infinite values in X_test :", test_inf)

Infinite values in X_train: 4281
Infinite values in X_test : 1090


In [4]:
# Train the Decision Tree model

print("Starting Decision Tree training...")

dt_model.fit(X_train, y_train)

print("Decision Tree training completed successfully.")

Starting Decision Tree training...


ValueError: Input X contains infinity or a value too large for dtype('float32').

In [2]:
from sklearn.tree import DecisionTreeClassifier

# Create Decision Tree model
dt_model = DecisionTreeClassifier(
    random_state=42
)

print("Decision Tree model created successfully.")
print(dt_model)

Decision Tree model created successfully.
DecisionTreeClassifier(random_state=42)


In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Load CIC-IDS2018 dataset
df = pd.read_csv("../cic.csv")

# Separate features and target
X = df.drop(columns=["Label"])
y = df["Label"]

# Select numeric features
X_numeric = X.select_dtypes(include="number")

# Create the common stratified train/test split
X_train, X_test, y_train, y_test = train_test_split(
    X_numeric,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("Training features:", X_train.shape)
print("Testing features :", X_test.shape)
print("Training target  :", y_train.shape)
print("Testing target   :", y_test.shape)

Training features: (838860, 78)
Testing features : (209715, 78)
Training target  : (838860,)
Testing target   : (209715,)


# Day 11 – Decision Tree Training

## Objective

Train the Decision Tree model using the common CIC-IDS2018 training input prepared during Days 8–10.

## Model

Decision Tree Classifier

## Training Input

- Training features: 78 numeric features
- Training rows: 838,860
- Target: `Label`
- Number of target classes: 3

## Evaluation

The trained model will be evaluated on the test set using:

- Accuracy
- Weighted Precision
- Weighted Recall
- Weighted F1 Score

A classification report will also be generated.

## Day 11 Scope

- Load the prepared dataset.
- Create the common train/test split.
- Train the Decision Tree model.
- Generate test predictions.
- Calculate evaluation metrics.
- Save the trained model.

Only the Decision Tree model will be trained on Day 11.